packages/paima-sdk/paima-providers/src/mina.ts

```diff
@@ -0,0 +1,169 @@
+    if (this.provider != null) {
+      return this.provider;
+    }
+
+    const options = MinaConnector.getWalletOptions();
+
+    if (options.length === 0) {
+      throw new WalletNotFound(`No Mina wallet found`);
+    }
+
+    return await this.connectExternal(gameInfo, await optionToActive(options[0]));
+  }
+
+  async connectNamed(gameInfo: GameInfo, name: string): Promise<IProvider<any>> {
+    if (this.provider?.getConnection().metadata?.name === name) {
+      return this.provider;
+    }
+    const provider = MinaConnector.getWalletOptions().find(entry => entry.metadata.name === name);
+    if (provider == null) {
+      throw new UnsupportedWallet(`MinaProvider: unsupported connection type ${name}`);
+    }
+    return await this.connectExternal(gameInfo, await optionToActive(provider));
+  }
+
+  connectExternal = async (
+    gameInfo: GameInfo,
+    conn: ActiveConnection<MinaApi>
+  ): Promise<MinaProvider> => {
+    if (this.provider?.getConnection().metadata?.name === conn.metadata.name) {
+      return this.provider;
+    }
+
+    this.provider = await MinaProvider.init(gameInfo, conn);
+    return this.provider;
+  };
+  getProvider = (): undefined | MinaProvider => {
+    return this.provider;
+  };
+  getOrThrowProvider = (): MinaProvider => {
+    if (this.provider == null) {
+      throw new ProviderNotInitialized(`MinaConnector provider isn't initialized yet`);
+    }
+    return this.provider;
+  };
+  isConnected = (): boolean => {
+    return this.provider != null;
+  };
+}
+
+export class MinaProvider implements IProvider<MinaApi> {
+  constructor(
+    private readonly conn: ActiveConnection<MinaApi>,
+    private readonly gameInfo: GameInfo,
+    readonly address: MinaAddress
+  ) {}
+
+  static init = async (
+    gameInfo: GameInfo,
+    conn: ActiveConnection<MinaApi>
+  ): Promise<MinaProvider> => {
+    const address = await MinaProvider.fetchAddress(conn);
+
+    return new MinaProvider(conn, gameInfo, address);
+  };
+  getConnection = (): ActiveConnection<MinaApi> => {
+    return this.conn;
+  };
+  getAddress = (): AddressAndType => {
+    return {
+      type: AddressType.MINA,
+      address: this.address,
+    };
+  };
+  signMessage = async (message: string): Promise<UserSignature> => {
+    // There is no way of choosing the signing account here. At most we could
+    // monitor the changed events and erroring out if it changed.
+    const signed = await this.conn.api.signMessage({ message: message });
+
+    if ('signature' in signed) {
+      const { signature } = signed;
+
+      const networkKind = await this.conn.api.requestNetwork();
+
+      // only these two options are used
+      // https://github.com/aurowallet/auro-wallet-browser-extension/blob/624de322dd99baaa09617bbad4a4a838f0a88edc/src/background/lib/index.js#L17
+      // this is part of the hashed message
+      const network = networkKind.chainId === 'mainnet' ? 'mainnet' : 'testnet';
```

Comment:
This is not true. Zeko is being added to Auro (there will be a canonical Zeko L2). Later on, there will probably be many Zeko-based networks in Auro - especially because for Paima we will probably end up where many games run as L3s

Reply:
Well... But it's still not the case right now, right? The problem is that the `signMessage` api doesn't tell you what's used for signing, and that's needed to verify. So to have this working for zeko I would need to know what will be the `chainId` in that case, and what will be the value used in there. Either that or the batcher needs to be pre-configured with the network id.

Reply:
Okay I asked about this and for now Zeko just uses "testnet" so it should work as-is for now with the code you have. I sent a message to the Auro team to see if they have any public documentation for how they plan to change these endpoints going forward, but it shouldn't block us for this PR. I created an issue to make sure we don't forget about this: https://github.com/PaimaStudios/paima-engine/issues/357